sound/soc/sof/ipc4-topology.c

```diff
@@ -1399,6 +1469,80 @@ static int sof_ipc4_prepare_src_module(struct snd_sof_widget *swidget,
 	return 0;
 }
 
+static int sof_ipc4_prepare_process_module(struct snd_sof_widget *swidget,
+					   struct snd_pcm_hw_params *fe_params,
+					   struct snd_sof_platform_stream_params *platform_params,
+					   struct snd_pcm_hw_params *pipeline_params, int dir)
+{
+	struct snd_soc_component *scomp = swidget->scomp;
+	struct snd_soc_dapm_widget *widget = swidget->widget;
+	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(scomp);
+	struct sof_ipc4_process *process = swidget->private;
+	struct snd_sof_control *scontrol = NULL;
+	void *extra_flag;
+	const struct snd_kcontrol_new *kc;
+	struct soc_bytes_ext *sbe;
+	struct sof_ipc4_control_data *control_data;
+	void *cfg = process->ipc_config_data;
+	int ret, i;
+
+	process->available_fmt.ref_audio_fmt = &process->available_fmt.base_config->audio_fmt;
+
+	/*
+	 * Output format is optional for process modules.
+	 * Process modules should handle the output format itself.
+	 */
+	ret = sof_ipc4_init_audio_fmt(sdev, swidget, &process->base_config,
+				      NULL, pipeline_params, &process->available_fmt,
+				      sizeof(process->base_config));
+	if (ret < 0)
+		return ret;
+
+	/* update pipeline memory usage */
+	sof_ipc4_update_pipeline_mem_usage(sdev, swidget, &process->base_config);
+
+	/*
+	 * 1. copy base_config to ipc_config_data
+	 * 2. copy output fmts to ipc_config_data
+	 * 3. copy extra payload to ipc_config_data
+	 */
+	/* update base_config */
+	memcpy(cfg, &process->base_config, sizeof(process->base_config));
+	cfg = cfg + sizeof(process->base_config);
```

Comment:
I think you should rename `cfg` to `ipc_payload`, because it is actually the payload we send to dsp.